src/acpustatus.cc

```diff
@@ -431,30 +481,82 @@ int CPUStatus::getAcpiTemp(char *tempbuf, int buflen) {
             if (strncmp(dir.entry(), "thermal", 7))
                 continue;
 
+            if (strlen(dir.entry()) >= MAXTZLEN) {
+                continue;
+            }
+
+            snprintf(namebuf, sizeof namebuf,
+                    "/sys/class/thermal/%s/type", dir.entry());
+            auto len = filereader(namebuf).read_all(BUFNSIZE(buf));
+            if (len) {
+                priorities[n_tzs] = tzPriority(buf);
+                strncpy(tzs[n_tzs++], dir.entry(), MAXTZLEN);
+            }
+            if (n_tzs >= MAXTEMPS) {
+                break;
+            }
+        }
+
+        // Sort thermal zones according to priority.
+        // inds are into tzs.
+        signed char inds[MAXTEMPS];
+        for (int i = 0; i < n_tzs; i++) inds[i] = i;
+        for (int i = 0; i < n_tzs; i++) {
+            signed char best = -1;
+            signed char bestind = -1;
+            for (int j = i; j < n_tzs; j++) {
+                if (bestind < 0 || priorities[j] > best) {
+                    best = priorities[j];
+                    bestind = j;
+                }
+            }
+
+            signed char temp = inds[i];
+            inds[i] = inds[bestind];
+            inds[bestind] = temp;
+            temp = priorities[i];
+            priorities[i] = best;
+            priorities[bestind] = temp;
+        }
+
+        for (int i = 0; i < n_tzs; i++) {
+            if (priorities[i] < 0) {
+                continue;
+            }
             snprintf(namebuf, sizeof namebuf,
-                    "/sys/class/thermal/%s/temp", dir.entry());
+                    "/sys/class/thermal/%s/temp", tzs[inds[i]]);
             auto len = filereader(namebuf).read_all(BUFNSIZE(buf));
             if (len > 4) {
+                // Remove the milligrades and newline.
                 int seglen = len - 4;
-                if (retbuflen + seglen + 4 >= buflen) {
+                int extra = longform ? 3 : 1;
+                if (needspace) extra++;
+                if (retbuflen + seglen + extra > buflen) {
```

Comment:
I changed the >= to a >, as there is already a 1 included for the \0, so retbuflen + seglen + extra can equal buflen, because only retbuflen + seglen + extra - 1 is used as index into buflen, unless I missed something.